core/feed_protocol/acquisition.py

```diff
@@ -0,0 +1,935 @@
+           selected facet.
+
+        :return: A dictionary of attributes, suitable for passing as
+            keyword arguments into OPDSFeed.add_link_to_feed.
+        """
+        args = dict(href=href, title=title)
+        args["rel"] = LinkRelations.FACET_REL
+        args["facetGroup"] = facet_group_name
+        if is_active:
+            args["activeFacet"] = "true"
+        return args
+
+    def as_error_response(self, **kwargs: Any) -> OPDSFeedResponse:
+        """Convert this feed into an OPDSFeedResponse that should be treated
+        by intermediaries as an error -- that is, treated as private
+        and not cached.
+        """
+        kwargs["max_age"] = 0
+        kwargs["private"] = True
+        return self.as_response(**kwargs)
+
+    @classmethod
+    def _create_entry(
+        cls,
+        work: Work,
+        active_licensepool: Optional[LicensePool],
+        edition: Edition,
+        identifier: Identifier,
+        annotator: Annotator,
+    ) -> WorkEntry:
+        entry = WorkEntry(
+            work=work,
+            license_pool=active_licensepool,
+            edition=edition,
+            identifier=identifier,
+        )
+        annotator.annotate_work_entry(entry)
+        return entry
+
+    ## OPDS1 specifics
+    @classmethod
+    def add_entrypoint_links(
+        cls,
+        feed: FeedData,
+        url_generator: Callable[[Type[EntryPoint]], str],
+        entrypoints: List[Type[EntryPoint]],
+        selected_entrypoint: Optional[Type[EntryPoint]],
+        group_name: str = "Formats",
+    ) -> None:
+        """Add links to a feed forming an OPDS facet group for a set of
+        EntryPoints.
+
+        :param feed: A FeedData object.
+        :param url_generator: A callable that returns the entry point
+            URL when passed an EntryPoint.
+        :param entrypoints: A list of all EntryPoints in the facet group.
+        :param selected_entrypoint: The current EntryPoint, if selected.
+        """
+        if len(entrypoints) == 1 and selected_entrypoint in (None, entrypoints[0]):
+            # There is only one entry point. Unless the currently
+            # selected entry point is somehow different, there's no
+            # need to put any links at all here -- a facet group with
+            # one one facet might as well not be there.
+            return
+
+        is_default = True
+        for entrypoint in entrypoints:
+            link = cls._entrypoint_link(
+                url_generator, entrypoint, selected_entrypoint, is_default, group_name
+            )
+            if link is not None:
+                feed.add_link(**link)
+                is_default = False
+
+    @classmethod
+    def _entrypoint_link(
+        cls,
+        url_generator: Callable[[Type[EntryPoint]], str],
+        entrypoint: Type[EntryPoint],
+        selected_entrypoint: Optional[Type[EntryPoint]],
+        is_default: bool,
+        group_name: str,
+    ) -> Optional[Dict[str, str]]:
```

Comment:
Your change here fixed this instance, but the `facet_link` function also returns a Dict instead of a link. It seems like it would make sense to follow this through, so we are passing around `Link` objects everywhere in this class instead of `Dict[str,str]`.